src/bedrock/monitor/policy_algorithm.cpp

```diff
@@ -0,0 +1,274 @@
+#include "spdlog/spdlog.h"
+#include "hash_ring.hpp"
+#include "monitor/monitoring_utils.hpp"
+
+using namespace std;
+
+void
+policy_algorithm(shared_ptr<spdlog::logger> logger,
+                 unordered_map<unsigned, GlobalHashRing>& global_hash_ring_map,
+                 unordered_map<unsigned, LocalHashRing>& local_hash_ring_map,
+                 chrono::time_point<chrono::system_clock>& grace_start,
+                 SummaryStats& ss,
+                 unsigned& adding_memory_node,
+                 unsigned& adding_ebs_node,
+                 bool& removing_memory_node,
+                 bool& removing_ebs_node,
+                 string management_address,
+                 unordered_map<string, KeyInfo>& placement,
+                 unordered_map<string, unsigned>& key_access_summary,
+                 MonitoringThread& mt,
+                 unordered_map<unsigned, TierData>& tier_data_map,
+                 unordered_map<string, unsigned>& departing_node_map,
+                 SocketCache& pushers,
+                 zmq::socket_t& response_puller,
+                 vector<string>& routing_address,
+                 unsigned& rid,
+                 unordered_map<string, pair<double, unsigned>>& rep_factor_map,
+                 const unsigned MINIMUM_MEMORY_NODE,
+                 const unsigned MINIMUM_EBS_NODE,
+                 const unsigned PROMOTE_THRESHOLD,
+                 const unsigned DEMOTE_THRESHOLD,
+                 const double MEM_CAPACITY_MAX,
+                 const double EBS_CAPACITY_MAX,
+                 const double EBS_CAPACITY_MIN,
+                 const unsigned VALUE_SIZE,
+                 const unsigned VIRTUAL_THREAD_NUM,
+                 const unsigned GRACE_PERIOD,
+                 const unsigned NODE_ADD
+                 ) {
+  unsigned memory_node_number = global_hash_ring_map[1].size() / VIRTUAL_THREAD_NUM;
+  unsigned ebs_node_number = global_hash_ring_map[2].size() / VIRTUAL_THREAD_NUM;
+
+  unordered_map<string, KeyInfo> requests;
+  unsigned total_rep_to_change = 0;
+
+  // 1. first check storage consumption and trigger elasticity if necessary
```

Comment:
Ideally, we'd break this algorithm up into many function calls. Your comments represent methods that could be checks (e.g., `storage consumption_limit_violated()`) and the actions (e.g., `trigger_elasticity`) would be mapped to them. You would then do something like `if (storage_consumption_limit_violated()) { trigger_elasticity(); }`. That will make our policies more composable (and readable).

Reply:
let me know how you feel about the structures in the latest commit

Reply:
Looks better! I think we should also separate actions from evaluation, but I think that is for another PR.

Reply:
Looks better! I think we should also separate actions from evaluation, but I think that is for another PR.